src/Analyser/TypeSpecifier.php

```diff
@@ -136,6 +138,16 @@ public function specifyTypesInCondition(
 				} else {
 					$type = new ObjectType($className);
 				}
+
+				$exprType = $scope->getType($exprNode);
+				if ($exprType instanceof GenericObjectType && $type instanceof TypeWithClassName) {
+					$type = new GenericObjectType(
+						$type->getClassName(),
+						$exprType->getTypes(),
```

Comment:
This is a bit too simple. It assumes the child class has the same number of `@template` tags and it uses them in the same order in `@extends`/`@implements`...

But all of these facts need to be checked beforehand before constructing GenericObjectType.

This problem is very similar to this one (so they should be solved at the same time) https://github.com/phpstan/phpstan/issues/2735